cardano-testnet/src/Testnet/Components/Query.hs

```diff
@@ -337,3 +342,54 @@ getCurrentEpochNo
 getCurrentEpochNo epochStateView = withFrozenCallStack $ do
   AnyNewEpochState _ newEpochState <- getEpochState epochStateView
   pure $ newEpochState ^. L.nesELL
+
+-- | Waits for a minimum of @minWait@ epochs and a maximum of @maxWait@ epochs for
+-- the value pointed by the @lens@ to become the same as the @mExpected@ (if it is not 'Nothing').
+-- If the value is not reached within the time frame, the test fails. If @mExpected@ is 'Nothing',
+-- the value is not checked, but the function will sitll wait for @minWait@ epochs.
+waitAndCheckNewEpochState
+  :: forall m era value.
+     (MonadAssertion m, MonadTest m, MonadIO m, Eq value)
+  => EpochStateView -- ^ Current epoch state view. It can be obtained using the 'getEpochStateView' function.
+  -> FilePath -- ^ The file path to the configuration file.
+  -> FilePath -- ^ The file path to the unix socket file to connect to the @cardano-node@.
+  -> ConwayEraOnwards era -- ^ Witness for the current era that shows it is Conway or onwards.
+  -> Word64 -- ^ The minimum wait time in epochs.
+  -> Maybe value -- ^ The expected value to check in the epoch state.
+  -> Word64 -- ^ The maximum wait time in epochs.
+  -> Lens' (L.NewEpochState (ShelleyLedgerEra era)) value -- ^ The lens to access the specific value in the epoch state.
+  -> m ()
+waitAndCheckNewEpochState epochStateView configurationFile socketPath ceo minWait mExpected maxWait lens = do
+  let sbe = conwayEraOnwardsToShelleyBasedEra ceo
+  (EpochNo curEpoch) <- getCurrentEpochNo epochStateView
+  eProposalResult
+    <- H.evalIO . runExceptT $ foldEpochState
+                                (File configurationFile)
+                                (File socketPath)
+                                FullValidation
+                                (EpochNo (curEpoch + maxWait))
+                                ()
+                                (\epochState _ _ -> filterEpochState (isSuccess curEpoch) epochState sbe)
+  void $ H.evalEither eProposalResult
+  where
+    filterEpochState :: (EpochNo -> value -> Bool) -> AnyNewEpochState -> ShelleyBasedEra era -> StateT () IO LedgerStateCondition
+    filterEpochState f (AnyNewEpochState actualEra newEpochState) sbe =
+        caseShelleyToBabbageOrConwayEraOnwards
+          (const $ error "waitAndCheck: Only conway era onwards supported")
+          (const $ do
+            Refl <- either error pure $ assertErasEqual sbe actualEra
+            let val = newEpochState ^. lens
+                currEpoch = L.nesEL newEpochState
+            return (if f currEpoch val
+                    then ConditionMet
+                    else ConditionNotMet)
+          )
+          sbe
+
+    isSuccess :: Word64 -> EpochNo -> value -> Bool
```

Comment:
What's the point of the minimum wait? What do we gain by knowing some value has been met in a particular range of epochs? It seems unnecessary.

Reply:
It is for use when we are checking a proposal didn't pass, the value is already there, so it would pass immediately. The minimum wait is there to ensure the value remains there for a while after the proposal was submitted (testing the proposal didn't pass). If the proposal is meant to pass and change something, then we set it to zero so that it waits as little as possible.

Reply:
> It is for use when we are checking a proposal didn't pass, the value is already there, so it would pass immediately.

Sorry I don't understand. Checking a proposal didn't pass but the value is already there? I'm lost. 

If the value is already there i.e we have:

```
(case mExpected of
        Nothing -> True
        Just expected -> value == expected) &&
      (epochNo <= epochAfterProp + maxWait)
```
evaluating to `True` so why do we care about the current epoch being more than the minimum wait?: `epochAfterProp + minWait <= epochNo`. We already have a termination epoch specified in `foldEpochState`, so why are we using it as a condition again in our callback?

Also shouldn't `Nothing` map to `False`?

Reply:
> Sorry I don't understand. Checking a proposal didn't pass but the value is already there? I'm lost.

For example, imagine we want to change the `drepActivity` parameter, and imagine it is currently:
```
drepActivity = 100
```
So we submit a proposal. Let's say we propose that we changange `drepActivity` so that `drepActivity = 101`. And we expect that proposal to fail for some reason (e.g: more stake voted `no` than voted `yes`). If the proposal indeed fails (as we expect), then the value after submitting and voting the proposal would still be:
```
drepActivity = 100
```
But this is not immediate, it takes an epoch at least for the change to have effect even if the proposal passes. So we need to wait a few epochs to ensure that `drepActivity` doesn't become `101` for a while, to ensure the proposal didn't pass. And that is where `minWait` comes in.

But if we were to expect the proposal to pass, we don't have to wait a few epochs, we just have to wait for the `drepActivity` to become `101`, the moment it does we know it worked, so we would set `minWait` to `0`.

Reply:
> Also shouldn't Nothing map to False as the expected value is not there and we don't want to terminate?

In this case, not having an expected value means we don't have an expectation, so whatever happens is fine with us. I do this when I need to fill epochs with proposals, but I am not certain whether enough DReps have expired yet for it to pass, so I have no expectation over those proposals.

Reply:
> But this is not immediate, it takes an epoch at least for the change to have effect even if the proposal passes. So we need to wait a few epochs to ensure that drepActivity doesn't become 101 for a while, to ensure the proposal didn't pass. And that is where minWait comes in.

Right this is clearer now. Split this into two functions please.

Reply:
Fixed here: https://github.com/IntersectMBO/cardano-node/pull/5815/commits/4a2444db1e5fa8f8cd2d7ff8fafaff527e0a116b